swift/extractor/trap/TrapLabel.h

```diff
@@ -38,7 +42,21 @@ class UntypedTrapLabel {
     return out;
   }
 
+  std::string str() const {
+    std::string ret(strSize(), '\0');
+    ret[0] = '#';
+    std::to_chars(ret.data() + 1, ret.data() + ret.size(), id_, 16);
+    return ret;
+  }
+
   friend bool operator==(UntypedTrapLabel lhs, UntypedTrapLabel rhs) { return lhs.id_ == rhs.id_; }
+
+ private:
+  size_t strSize() const {
+    if (id_ == undefined) return 17;  // #ffffffffffffffff
+    if (id_ == 0) return 2;           // #0
+    return /* # */ 1 + /* hex digits */ static_cast<size_t>(ceil(log2(id_ + 1) / 4));
```

Comment:
Optional: The floating-point log looks painful. Are any of the following better options?
 1. C++20 `std::bit_width`;
 2. Abseil's backport `absl::bit_width`;
 3. A `while (id >>= 4) digits+=4;`-type loop (which unfortunately compiles to a literal shift/add/cmp loop instead of using LZCNT);
 4. Return a fixed-size, null-terminated buffer from `str()` (e.g. `std::array<char, 18>`) and use `strlen(str().data())`.

Reply:
via the `mserialize` specialisations this is also used for every log. That's why I wanted to avoid option 4, to avoid each label being converted twice for each log (once to calculate its size, another to store it). Maybe I'll leave a todo to switch to `bit_width`, whichever comes first between abseil and C++20 :slightly_smiling_face:

Reply:
I did say "optional", so a todo sounds good. That said, I wouldn't make any bets on a floating-point log being faster than an int->hex conversion!

Reply:
Yeah, good point. But I think I'll leave it like this with a TODO, as it makes for a good hook for inserting `bit_width` later on.